### DISCUSSION TITLE
Brodcasting single message to all subscribers

### DISCUSSION THREAD

**Discussion by mentos1386** [enisdenjo/graphql-ws#423] [Ideas]
Hi!

I'm thininking of an usecase where clients can subscribe to `heartbeat` like this:
```graphql
subscription {
  heartbeat {
     foo
  }
}
```

We would like to all subscribers to send the exactly the same `heartbeat` objects. There is no problems that we would be "leaking" information as this is a subscription that is intentional public and available for everyone with content that is the same for everyone.

What I'm afraid is, if we have 100 subscribers and send a single message, we will be processing (calling "resolve" functions) 100 times for each subscriber. Even though the outcome for each subscriber will be the same.

We want to avoid getting our "resolve" methods being called 100 times if they could only be called once.

Maybe in the way [graphql-ruby](https://graphql-ruby.org/subscriptions/broadcast.html) mentions?

**Comment by enisdenjo** [comment]
Hey there, this is more of a general GraphQL issue rather than graphql-ws specific. graphql-ws is just a transport.

You can create a pubsub that is globally shared for the `heartbeat` field and publish to that pubsub to notify all subscribed clients, if you deem invoking resolve as expensive - you can memoize it.

**Reply by mentos1386** [reply]
> Hey there, this is more of a general GraphQL issue rather than graphql-ws specific. graphql-ws is just a transport.

I guess that makes sense.

> You can create a pubsub that is globally shared for the heartbeat field and publish to that pubsub to notify all subscribed clients

What do you mean with `globally shared`? I don't understand what you mean.

> if you deem invoking resolve as expensive - you can memoize it.

This is our solution. We have used [dataloader](https://github.com/graphql/dataloader) where the same dataloader is shared for all "subscribers" so the "resolve" is being called N times but the dataloader only produces 1 "resolve".

**Reply by enisdenjo** [reply]
> What do you mean with `globally shared`? I don't understand what you mean.

Sorry oversaw your comment - I meant a pubsub singleton that each publisher uses to publish and every listener to subscribe.
